src/devices/cpu/arcompact/arcompactdasm_ops_04.cpp

```diff
@@ -0,0 +1,482 @@
+} // *
+
+int arcompact_disassembler::handle_dasm32_MPYH(std::ostream &stream, offs_t pc, uint32_t op, const data_buffer &opcodes)
+{
+	return handle04_helper_dasm(stream, pc, op, opcodes, "MPYH", 0,0);
+} // *
+
+int arcompact_disassembler::handle_dasm32_MPYHU(std::ostream &stream, offs_t pc, uint32_t op, const data_buffer &opcodes)
+{
+	return handle04_helper_dasm(stream, pc, op, opcodes, "MPYHU", 0,0);
+} // *
+
+int arcompact_disassembler::handle_dasm32_MPYU(std::ostream &stream, offs_t pc, uint32_t op, const data_buffer &opcodes)
+{
+	return handle04_helper_dasm(stream, pc, op, opcodes, "MPYU", 0,0);
+} // *
+
+int arcompact_disassembler::handle_dasm32_Jcc(std::ostream &stream, offs_t pc, uint32_t op, const data_buffer &opcodes)
+{
+	return handle04_helper_dasm(stream, pc, op, opcodes, "J", 1,1);
+}
+
+int arcompact_disassembler::handle_dasm32_Jcc_D(std::ostream &stream, offs_t pc, uint32_t op, const data_buffer &opcodes)
+{
+	return handle04_helper_dasm(stream, pc, op, opcodes, "J.D", 1,1);
+}
+
+int arcompact_disassembler::handle_dasm32_JLcc(std::ostream &stream, offs_t pc, uint32_t op, const data_buffer &opcodes)
+{
+	return handle04_helper_dasm(stream, pc, op, opcodes, "JL", 1,1);
+}
+
+int arcompact_disassembler::handle_dasm32_JLcc_D(std::ostream &stream, offs_t pc, uint32_t op, const data_buffer &opcodes)
+{
+	return handle04_helper_dasm(stream, pc, op, opcodes, "JL.D", 1,1);
+}
+
+int arcompact_disassembler::handle_dasm32_LP(std::ostream& stream, offs_t pc, uint32_t op, const data_buffer& opcodes) // LPcc (loop setup)
+{
+	uint8_t breg = dasm_common32_get_breg(op); // breg is reserved
+	uint8_t p = dasm_common32_get_p(op);
+
+	if (p == 0x00)
+	{
+		util::stream_format(stream, "<illegal LPcc, p = 0x00)");
+	}
+	else if (p == 0x01)
+	{
+		util::stream_format(stream, "<illegal LPcc, p = 0x01)");
+	}
+	else if (p == 0x02) // Loop unconditional
+	{ // 0010 0RRR 1010 1000 0RRR ssss ssSS SSSS
+		uint32_t S = dasm_common32_get_s12(op);
+		if (S & 0x800) S = -0x800 + (S & 0x7ff);
+
+		util::stream_format(stream, "LP (start %08x, end %08x)", pc + 4, (pc & 0xfffffffc) + S * 2);
+	}
+	else if (p == 0x03) // Loop conditional
+	{ // 0010 0RRR 1110 1000 0RRR uuuu uu1Q QQQQ
+		uint32_t u = dasm_common32_get_u6(op);
+		uint8_t condition = dasm_common32_get_condition(op);
+		util::stream_format(stream, "LP<%s> (start %08x, end %08x)", conditions[condition], pc + 4, (pc & 0xfffffffc) + u * 2);
+
+		int unused = (op & 0x00000020) >> 5;
+		if (unused == 0)  util::stream_format(stream, "(unused bit not set)");
+
+	}
+
+	if (breg) util::stream_format(stream, "(reseved B bits set %02x)", breg);
+
+	return 4;
+}
+
+#define PRINT_AUX_REGNAME \
+	if ((auxreg >= 0) && (auxreg < 0x420)) \
+	{ \
+		if (strcmp(auxregnames[auxreg],"unusedreg")) \
+			util::stream_format(stream, "[%s]", auxregnames[auxreg]); \
+		else \
+			util::stream_format(stream, "[%03x]", auxreg); \
+	} \
+	else \
+		util::stream_format(stream, "[%03x]", auxreg);
```

Comment:
The disassembler code is mostly what it was, aside from some small cleanups to fix disassembly problems, properly cleaning it up was my next step after this lot goes in.

I don't think I've made any of the disassembler stuff worse, just moved things around so that it will be easier to work with.  You're commenting on code that is already in MAME here.

Reply:
cleaned it up for now anyway